internal/rbd/controllerserver.go

```diff
@@ -271,19 +271,27 @@ func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSna
 func (cs *ControllerServer) CreateVolume(
 	ctx context.Context,
 	req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	if err := cs.validateVolumeReq(ctx, req); err != nil {
+	var err error
+	if err = cs.validateVolumeReq(ctx, req); err != nil {
 		return nil, err
 	}
 
 	// TODO: create/get a connection from the the ConnPool, and do not pass
 	// the credentials to any of the utility functions.
-	cr, err := util.NewUserCredentials(req.GetSecrets())
+	secrets := req.GetSecrets()
+	if util.IsMigrationSecret(secrets) {
+		secrets, err = util.ParseAndSetSecretMapFromMigSecret(secrets)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+	cr, err := util.NewUserCredentials(secrets)
```

Comment:
why this change is needed?

Reply:
this is required when we added migration secret support https://github.com/ceph/ceph-csi/pull/2562 , otherwise validation of the request will fail. @Madhu-1

Reply:
why you are adding it to createVolume? the migration is only to support the existing volume, isn't it?

Reply:
I am not adding anything here @Madhu-1, not sure what is the confusion here. Migration is for existing intree storage classes. As pointed out in all the migration documentations it is the very basic thing expected from migration,  for ex: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md
```
Dynamically Provisioned volumes will continue to be provisioned with the in-tree PersistentVolumeSource. The CSI external-provisioner to pick up the in-tree PVC's when migration is turned on and provision using the CSI Drivers;...
```

Reply:
Do you mean to say that the user will still create the CSI volumes using storageclass which contains the intree details?

Reply:
it is upto the admin , mostly what I heard was that, new CSI SCs will be given to the users thus new volume on CSI SC as anyway they are in migration path.

Reply:
IMHO we should not support this one in cephcsi and just stick to the plan to treat migration as static only add documentation to use new SC for csi volumes.

Reply:
This is very well understood and clear that new volumes has to go through and its a must requirement. We were always in plan and  never deviated from it . Also, this was  discussed in  CSI calls. `Static` treatment is only applicable to the existing intree volumes.

Reply:
@humblec can you please share with me the link to a  chat/doc where this is discussed in csi call? at least am not aware of any discussion. for migration, we are getting everything in bit and piece and this was not part of your design doc. now, this is news for me, and am not aware of it.

Reply:
>
for migration, we are getting everything in bit and piece and this was not part of your design doc. now, this is news for me, and am not aware of it.
>
@Madhu-1 it was discussed  in csi stand up calls, may be you were unavailable on those calls. It was/is part of the tracker issue, design doc ( with above link in place)..etc and there was also an explicit  mention of  provisioning and other operations in the design doc which was removed when you asked to replace that section by this comment https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 . That section was replaced by the review comment  ask as it was repeating the same thing.. It is really surprising that,  it is a news . Also, this is not breaking change at all!! All existing functionality works and helping users if they opted for migration.

Reply:
@nixpanic are you aware of this discussion?

Reply:
@Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this..Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!!

Reply:
@humblec at least I was not aware of any of these till I review this PR. the reason for asking to remove the duplicate content from the design doc here is that https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 does not want to have the same exact copy-paste from the kubernetes design doc in cephcsi. if we get the design PR in first and then the implementation we can avoid these discussions and also we are doing multiple PR to achieve this one we can avoid doing cleanup introduced in the last PR like we are doing in this PR hope you got my concern

Reply:
@Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this.

I don't remember discussing create or resize of volume using intree storage class at that point in time.

>.Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!! 


IMO, We should be encouraging the usage of csi storageclass.

Reply:
>@Madhu-1 *at least* I remember discussing these as parf of these dates `2021-09-13 , 2021-09-14` in CSI standup calls. https://docs.google.com/document/d/1K1aerdMpraIh56-skdoEoVF9RZrO4NUcbHtjN-f3u1s/edit#heading=h.z6gjxa34z215 .  Atleast @Rakshith-R should be aware of this.
> 
> I don't remember discussing create or resize of volume using intree storage class at that point in time.
> 
> >.Also I have provided updates in other stand up calls too. More or less, I still didnt get what is the `real concern` here? it is not breaking any existing functionality and rather helping users !!! 
> 
> 
> IMO, We should be encouraging the usage of csi storageclass.
> 
> 
> 
> 
>  
> 
>

Reply:
>@humblec at least I was not aware of any of these till I review this PR. the reason for asking to remove the duplicate content from the design doc here is that https://github.com/ceph/ceph-csi/pull/2609/files#r744505983 does not want to have the same exact copy-paste from the kubernetes design doc in cephcsi. if we get the design PR in first and then the implementation we can avoid these discussions and also we are doing multiple PR to achieve this one we can avoid doing cleanup introduced in the last PR like we are doing in this PR hope you got my concern
>

@Madhu-1 nw at all, I am glad that you asked the clarification. once again I was also surprised by reading this review comment. We can definitely make the process better if there are flaws , end of the day its all about helping users with the project.  :+1:

Reply:
'I don't remember discussing create or resize of volume using intree storage class at that point in time.' , 
@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).

Reply:
@humblec IMHO design discussion should include the maintainers :)

Reply:
IMO the whole process is a complete mess for me at least. if we don't have proper path things become hard at least to understand the goal and to make decisions.

Reply:
@Madhu-1 I can take an action item to have a discussion among us and document the process here in the repo from next release onwards.. :+1:

Reply:
> .@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).

@humblec  As far as i remember we mainly discussed mapping required for vol-id changes. 

I only supported the node operations for older intree pvc as they are crucial,
I still don't feel we should support other functions and instead should encourage users to switch to **csi storageclass** going forward. :)

Reply:
> > .@Rakshith-R please refresh :), resize came in late, but new volumes and annotation hack was discussed :).
> 
> @humblec As far as i remember we mainly discussed mapping required for vol-id changes.
> 
the vol id part  was for existing volumes, in that call new vols and annotations to identify or distinguish the volid were also discussed. any way there is no point of going back :)

Reply:
The design doc should be updated to include what changes are specific for Ceph-CSI. That includes mentioning functions like `ParseAndSetSecretMapFromMigSecret()`, and when that needs to be used.

For migration, I was expecting that the old in-tree PVs would be migrated to new CSI PVs. But it seems there is no real migration of that  kind. It currently looks like in-tree PVs are not _migrated_, but rather support for the in-tree format is added to Ceph-CSI. This is a little confusing...

The design doc should probably describe in a little more detail what CSI procedures need support for the old in-tree PVs for _migration_, and how that functionality is added to Ceph-CSI. A separate PR for the (updated) design doc might be appropriate, and that should be approved/merged before implementing the needed changes (ideally).

Reply:
Added https://github.com/ceph/ceph-csi/pull/2679 ..